core/src/main/kotlin/io/github/kkafka/producer/transaction.kt

```diff
@@ -0,0 +1,91 @@
+ * The [consumerTopics] indicate which of the consumer's topics are taking part in the transaction.
+ * You should not include the producer's topic.
+ * The [consumer] must be subscribed to and consuming from the given [consumerTopics].
+ *
+ * [block] is the transaction code. Here you can consume from [consumer] and send with the producer.
+ * At the end of the [block], if no exceptions were thrown, the transaction is committed and the result of the
+ * block is returned. Otherwise, the transaction is aborted and the exception is thrown.
+ *
+ * This is useful for a consume-process-produce pattern, where we commit to the consumer only if we have successfully
+ * produced. This ensures exactly once processing.
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(
+    consumer: KafkaConsumer<*, *>,
+    consumerTopics: Set<String>,
+    block: KafkaProducer<K, V>.() -> T,
+): T {
+    beginTransaction()
+    return try {
+        val res = block(this)
+
+        // Get the TopicPartition objects given the topic names
+        val consumerTopicPartitions = consumer.assignment().filter { it.topic() in consumerTopics }.map { it!! }
+
+        // Get the offsets of the partitions so we can commit them as part of the transaction
+        val offsets = consumerTopicPartitions.associateWith { OffsetAndMetadata(consumer.position(it)) }
+
+        val groupId = consumer.groupMetadata().groupId()
+
+        // Make sure that we got all the needed partitions
+        if (offsets.size != consumerTopics.size) {
+            throw TransactionAbortedException(
+                "consumer in transaction is not properly subscribed to all topics given for transaction"
+            )
+        }
+
+        // Commit the consumer's offsets
+        sendOffsetsToTransaction(offsets, groupId)
+
+        // Successful transaction
+        commitTransaction()
+
+        res
+    } catch (e: Exception) {
+        // If the transaction failed, abort
+        abortTransaction()
+        throw e
+    }
+}
+
+/**
+ * Consume and produce transactionally on a single topic.
+ * See [transaction].
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(
+    consumer: KafkaConsumer<*, *>,
+    consumerTopic: String,
+    block: KafkaProducer<K, V>.() -> T,
+): T {
+    return try {
+        transaction(consumer, setOf(consumerTopic), block)
+    } catch (e: Exception) {
+        throw e
+    }
+}
+
+/**
+ * Produce transactionally with an arbitrary block of code.
+ * If an exception occurs during execution of [block], the transaction is aborted, otherwise it is committed
+ * and the result of [block] is returned.
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(block: KafkaProducer<K, V>.() -> T): T {
+    beginTransaction()
+    return try {
+        val res = block(this)
+        commitTransaction()
+        res
+    } catch (e: Exception) {
+        abortTransaction()
```

Comment:
is it correct to call abort() if it was commit() that threw?

Reply:
(I am worried abort() may throw)

Reply:
"If any of the send calls failed with an irrecoverable error, the final commitTransaction() call will fail and throw the exception from the last failed send. When this happens, your application should call abortTransaction() to reset the state and continue to send data."

- https://kafka.apache.org/10/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send--

that's what expected, except in the case of the 4 exceptions listed later on for which the producer should be closed. I've implemented that in the last commit.